lib/ModuleFinder.js

```diff
@@ -0,0 +1,216 @@
+import path from 'path';
+
+import logLevel from 'loglevel';
+
+import ExportsStorage from './ExportsStorage';
+import Watcher from './Watcher';
+import findExports from './findExports';
+import findPackageDependencies from './findPackageDependencies';
+import normalizePath from './normalizePath';
+import readFile from './readFile';
+import requireResolve from './requireResolve';
+
+/**
+ * Checks for package.json or npm-shrinkwrap.json inside a list of files and
+ * expands the list of files to include package dependencies if so.
+ */
+function expandFiles(files, workingDirectory) {
+  return new Promise((resolve) => {
+    const result = new Set([]);
+    files.forEach((file) => {
+      if (file.path !== './package.json' &&
+        file.path !== './npm-shrinkwrap.json') {
+        result.add(file);
+        return;
+      }
+      findPackageDependencies(workingDirectory, true).forEach((dep) => {
+        const pathToResolve = path.join(
+          workingDirectory,
+          'node_modules',
+          dep
+        );
+        const resolvedPath = requireResolve(pathToResolve);
+        if (resolvedPath === pathToResolve) {
+          // Ignore
+          return;
+        }
+
+        result.add({
+          path: normalizePath(resolvedPath, workingDirectory),
+          mtime: file.mtime,
+          alias: dep,
+        });
+      });
+    });
+    resolve([...result]);
+  });
+}
+
+function aliasedExportNames(alias, ignorePackagePrefixes) {
+  const result = [alias];
+  ignorePackagePrefixes.forEach((prefix) => {
+    if (alias.startsWith(prefix)) {
+      result.push(alias.slice(prefix.length));
+    }
+  });
+  return result;
+}
+
+function defaultExportNames(pathToFile) {
+  const parsed = path.parse(pathToFile);
+  let fileName = parsed.name;
+  let dirName = path.basename(parsed.dir);
+
+  if (/package\.json|index\.jsx?/.test(parsed.base)) {
+    fileName = dirName;
+    dirName = path.basename(path.dirname(parsed.dir));
+  }
+  const result = [
+    `${dirName}-${fileName}`,
+    fileName,
+  ];
+
+  if (/e?s$/.test(dirName)) {
+    result.push(`${dirName.replace(/e?s$/, '')}-${fileName}`);
+  }
+  return result;
+}
+
+const instances = {};
+
+export default class ModuleFinder {
+  /**
+   * Factory method to get an instance for a specific working directory.
+   */
+  static getForWorkingDirectory(workingDirectory, { excludes, ignorePackagePrefixes }) {
+    let instance = instances[workingDirectory];
+    if (!instance) {
+      instance = new ModuleFinder(workingDirectory, {
+        excludes,
+        ignorePackagePrefixes,
+      });
+      instances[workingDirectory] = instance;
+    }
+    return instance;
+  }
+
+  constructor(workingDirectory, { excludes, ignorePackagePrefixes }) {
+    this.excludes = excludes;
```

Comment:
Yeah. If you can think of ways to make this better, drop a line or so here. One of the reasons state is needed is because we have long-running instances, and the call sites usually don't have enough state themselves to pass in the right things.